File: app/src/main/java/com/example/despensacx/ProductoAdapter.java
package com.example.despensacx;

import android.graphics.Color;
import android.graphics.Paint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.ImageButton;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class ProductoAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_HEADER = 0;
    private static final int TYPE_ITEM = 1;

    public static class DisplayItem {
        public int type;
        public String tiendaNombre;
        public String tiendaColor;
        public double subtotalTienda;
        public ProductoConTienda productoConTienda;

        public DisplayItem(String tiendaNombre, String tiendaColor, double subtotalTienda) {
            this.type = TYPE_HEADER;
            this.tiendaNombre = tiendaNombre;
            this.tiendaColor = tiendaColor;
            this.subtotalTienda = subtotalTienda;
        }

        public DisplayItem(ProductoConTienda productoConTienda) {
            this.type = TYPE_ITEM;
            this.productoConTienda = productoConTienda;
        }
    }

    public interface OnProductoActionListener {
        void onToggleSeleccion(ProductoEntity producto, boolean seleccionado);
        void onEditar(ProductoEntity producto);
        void onEliminar(ProductoEntity producto);
    }

    private List<DisplayItem> items = new ArrayList<>();
    private final OnProductoActionListener listener;

    public ProductoAdapter(OnProductoActionListener listener) {
        this.listener = listener;
    }

    public void setItems(List<DisplayItem> items) {
        this.items = items;
        notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
        return items.get(position).type;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == TYPE_HEADER) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_header_tienda, parent, false);
            return new HeaderViewHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_producto, parent, false);
            return new ProductoViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        DisplayItem item = items.get(position);
        if (holder instanceof HeaderViewHolder) {
            ((HeaderViewHolder) holder).bind(item);
        } else if (holder instanceof ProductoViewHolder) {
            ((ProductoViewHolder) holder).bind(item.productoConTienda, listener);
        }
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    static class HeaderViewHolder extends RecyclerView.ViewHolder {
        TextView tvNombreTienda, tvSubtotalTienda;
        View viewIndicatorColor;

        public HeaderViewHolder(@NonNull View itemView) {
            super(itemView);
            tvNombreTienda = itemView.findViewById(R.id.tvHeaderNombreTienda);
            tvSubtotalTienda = itemView.findViewById(R.id.tvHeaderSubtotal);
            viewIndicatorColor = itemView.findViewById(R.id.viewHeaderColor);
        }

        public void bind(DisplayItem item) {
            tvNombreTienda.setText(item.tiendaNombre);
            tvSubtotalTienda.setText(String.format(Locale.getDefault(), "Subtotal: $%.2f", item.subtotalTienda));
            try {
                viewIndicatorColor.setBackgroundColor(Color.parseColor(item.tiendaColor));
            } catch (Exception e) {
                viewIndicatorColor.setBackgroundColor(Color.GRAY);
            }
        }
    }

    static class ProductoViewHolder extends RecyclerView.ViewHolder {
        CheckBox cbSeleccionado;
        TextView tvDescripcion, tvDetallePrecio, tvImporteTotal;
        ImageButton btnEditar, btnEliminar;

        public ProductoViewHolder(@NonNull View itemView) {
            super(itemView);
            cbSeleccionado = itemView.findViewById(R.id.cbProducto);
            tvDescripcion = itemView.findViewById(R.id.tvDescripcionProducto);
            tvDetallePrecio = itemView.findViewById(R.id.tvDetallePrecio);
            tvImporteTotal = itemView.findViewById(R.id.tvImporteTotal);
            btnEditar = itemView.findViewById(R.id.btnEditarProducto);
            btnEliminar = itemView.findViewById(R.id.btnEliminarProducto);
        }

        public void bind(ProductoConTienda pct, OnProductoActionListener listener) {
            ProductoEntity p = pct.getProducto();
            tvDescripcion.setText(p.getCantidad() + "x " + p.getDescripcion());
            tvDetallePrecio.setText(String.format(Locale.getDefault(), "P.U.: $%.2f", p.getPrecio()));

            double total = p.getPrecio() * p.getCantidad();
            tvImporteTotal.setText(String.format(Locale.getDefault(), "$%.2f", total));

            // Evitar disparar listener al rebind
            cbSeleccionado.setOnCheckedChangeListener(null);
            cbSeleccionado.setChecked(p.isSeleccionado());

            if (p.isSeleccionado()) {
                tvDescripcion.setPaintFlags(tvDescripcion.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG));
                tvImporteTotal.setPaintFlags(tvImporteTotal.getPaintFlags() & (~Paint.STRIKE_THRU_TEXT_FLAG));
                itemView.setAlpha(1.0f);
            } else {
                tvDescripcion.setPaintFlags(tvDescripcion.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
                tvImporteTotal.setPaintFlags(tvImporteTotal.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
                itemView.setAlpha(0.5f);
            }

            cbSeleccionado.setOnCheckedChangeListener((buttonView, isChecked) -> {
                listener.onToggleSeleccion(p, isChecked);
            });

            btnEditar.setOnClickListener(v -> listener.onEditar(p));
            btnEliminar.setOnClickListener(v -> listener.onEliminar(p));
        }
    }
}

File: app/src/main/kotlin/com/example/despensacx/adapter/ProductoAdapter.kt
package com.example.despensacx.adapter

import android.graphics.Color
import android.graphics.Paint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageButton
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.despensacx.R
import com.example.despensacx.data.ProductoEntity
import com.example.despensacx.model.ProductoConTienda
import com.example.despensacx.utils.FormatUtils
import java.util.*

class ProductoAdapter(private val listener: OnProductoActionListener) :
    RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val TYPE_HEADER = 0
        private const val TYPE_ITEM = 1
    }

    data class DisplayItem(
        val type: Int,
        val tiendaNombre: String? = null,
        val tiendaColor: String? = null,
        val subtotalTienda: Double = 0.0,
        val productoConTienda: ProductoConTienda? = null
    ) {
        constructor(tiendaNombre: String, tiendaColor: String, subtotalTienda: Double) : 
            this(TYPE_HEADER, tiendaNombre, tiendaColor, subtotalTienda)
        
        constructor(productoConTienda: ProductoConTienda) : 
            this(TYPE_ITEM, productoConTienda = productoConTienda)
    }

    interface OnProductoActionListener {
        fun onToggleSeleccion(producto: ProductoEntity, seleccionado: Boolean)
        fun onEditar(producto: ProductoEntity)
        fun onEliminar(producto: ProductoEntity)
    }

    private var items: List<DisplayItem> = emptyList()

    fun setItems(items: List<DisplayItem>) {
        this.items = items
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int = items[position].type

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == TYPE_HEADER) {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_header_tienda, parent, false)
            HeaderViewHolder(view)
        } else {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_producto, parent, false)
            ProductoViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = items[position]
        when (holder) {
            is HeaderViewHolder -> holder.bind(item)
            is ProductoViewHolder -> item.productoConTienda?.let { holder.bind(it, listener) }
        }
    }

    override fun getItemCount(): Int = items.size

    class HeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val tvNombreTienda: TextView = itemView.findViewById(R.id.tvHeaderNombreTienda)
        private val tvSubtotalTienda: TextView = itemView.findViewById(R.id.tvHeaderSubtotal)
        private val viewIndicatorColor: View = itemView.findViewById(R.id.viewHeaderColor)

        fun bind(item: DisplayItem) {
            tvNombreTienda.text = item.tiendaNombre
            tvSubtotalTienda.text = itemView.context.getString(R.string.subtotal_tienda, FormatUtils.formatCurrency(item.subtotalTienda))
            try {
                viewIndicatorColor.setBackgroundColor(Color.parseColor(item.tiendaColor))
            } catch (e: Exception) {
                viewIndicatorColor.setBackgroundColor(Color.GRAY)
            }
        }
    }

    class ProductoViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val cbSeleccionado: CheckBox = itemView.findViewById(R.id.cbProducto)
        private val tvDescripcion: TextView = itemView.findViewById(R.id.tvDescripcionProducto)
        private val tvDetallePrecio: TextView = itemView.findViewById(R.id.tvDetallePrecio)
        private val tvImporteTotal: TextView = itemView.findViewById(R.id.tvImporteTotal)
        private val btnEditar: ImageButton = itemView.findViewById(R.id.btnEditarProducto)
        private val btnEliminar: ImageButton = itemView.findViewById(R.id.btnEliminarProducto)

        fun bind(pct: ProductoConTienda, listener: OnProductoActionListener) {
            val p = pct.producto
            tvDescripcion.text = itemView.context.getString(R.string.cantidad_descripcion, p.cantidad, p.descripcion)
            tvDetallePrecio.text = String.format(Locale.getDefault(), "P.U.: $%,.2f", p.precio)

            val total = p.precio * p.cantidad
            tvImporteTotal.text = String.format(Locale.getDefault(), "$%,.2f", total)

            cbSeleccionado.setOnCheckedChangeListener(null)
            cbSeleccionado.isChecked = p.seleccionado

            if (p.seleccionado) {
                tvDescripcion.paintFlags = tvDescripcion.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
                tvImporteTotal.paintFlags = tvImporteTotal.paintFlags and Paint.STRIKE_THRU_TEXT_FLAG.inv()
                itemView.alpha = 1.0f
            } else {
                tvDescripcion.paintFlags = tvDescripcion.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
                tvImporteTotal.paintFlags = tvImporteTotal.paintFlags or Paint.STRIKE_THRU_TEXT_FLAG
                itemView.alpha = 0.5f
            }

            cbSeleccionado.setOnCheckedChangeListener { _, isChecked ->
                listener.onToggleSeleccion(p, isChecked)
            }

            btnEditar.setOnClickListener { listener.onEditar(p) }
            btnEliminar.setOnClickListener { listener.onEliminar(p) }
        }
    }
}
